File: examples/footguns/03_memory_leak/main.c
/*
 * Footgun: Memory Leak — C
 *
 * An early-return on the error path skips the free(). The allocation
 * is reachable nowhere after the function returns, but it is never
 * freed. C has no destructors, no defer, no static analysis built in.
 * The leak is silent unless you run valgrind or ASan.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUFSIZE 256

/* Returns a heap-allocated processed string, or NULL on error.
 * BUG: the buf allocation is leaked when input starts with '!'. */
char *process(const char *input) {
    char *buf = malloc(BUFSIZE);
    if (!buf) return NULL;

    snprintf(buf, BUFSIZE, "processed: %s", input);

    if (input[0] == '!') {
        /* Error path: forgot to free(buf) before returning NULL. */
        return NULL; /* leak: buf is gone, never freed */
    }

    return buf; /* caller must free this */
}

int main(void) {
    /* Happy path: result is freed correctly. */
    char *r1 = process("hello");
    if (r1) { printf("%s\n", r1); free(r1); }

    /* Error path: buf inside process() is leaked silently. */
    char *r2 = process("!bad");
    if (!r2) printf("error (and a leak just happened)\n");

    return 0;
}

/*
 * Compile and run:
 *   gcc -o leak main.c && ./leak
 *   valgrind --leak-check=full ./leak
 *   gcc -fsanitize=address -o leak main.c && ./leak
 *
 * Result without tools: silent leak.
 * Result with valgrind/ASan: "definitely lost: 256 bytes in 1 block"
 */

File: examples/footguns/03_memory_leak/main.rs
// Footgun: Memory Leak — Rust
//
// Rust's RAII (Drop trait) makes the common C leak pattern impossible:
// every heap allocation is tied to an owner, and the owner's destructor
// runs unconditionally when the owner goes out of scope — including on
// early returns and panics. No manual free() to forget.
//
// There are two intentional escape hatches that CAN leak:
//   mem::forget(v)  — explicitly suppresses Drop
//   Box::leak(b)    — intentionally converts a Box into a &'static ref
// Both require you to explicitly opt in. Accidental leaks via forgotten
// free() are not a category in safe Rust.

fn process(input: &str) -> Option<String> {
    let buf = format!("processed: {}", input); // heap allocation

    if input.starts_with('!') {
        return None; // buf is dropped here automatically — no leak
    }

    Some(buf) // ownership moves to caller
}

fn main() {
    // Happy path
    if let Some(r) = process("hello") {
        println!("{}", r);
    } // r dropped here

    // Error path: buf inside process() is dropped at the early return.
    // There is no leak.
    if process("!bad").is_none() {
        println!("error (no leak: Rust dropped buf at the early return)");
    }

    // -----------------------------------------------------------------
    // Intentional leak via mem::forget — requires explicit opt-in:
    let v = vec![1, 2, 3];
    std::mem::forget(v); // explicitly suppress Drop; v's memory leaks
    println!("intentional leak via mem::forget (requires explicit call)");

    // Box::leak for 'static references — also intentional:
    let s: &'static str = Box::leak(Box::new(String::from("static")));
    println!("leaked static ref: {}", s);
}

// Compile and run:  rustc main.rs -o leak && ./leak
// Valgrind will show the mem::forget and Box::leak allocations, which
// are intentional. The process() early-return path is leak-free.
